File: src/shared/game/voxel/StructurePlacer.ts
import { VoxelChunk, CHUNK_SIZE_X, CHUNK_SIZE_Y, CHUNK_SIZE_Z } from './VoxelChunk';
import { VOXEL_WORD_AIR_LOW } from './VoxelWord';

/** Fast deterministic PRNG (mulberry32) */
function mulberry32(a: number) {
  return function() {
    var t = a += 0x6D2B79F5;
    t = Math.imul(t ^ t >>> 15, t | 1);
    t ^= t + Math.imul(t ^ t >>> 7, t | 61);
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  }
}

export interface ProceduralStructure {
  voxelMapId: string;
  spawnWeight: number; // 0.0 to 1.0
  yOffset: number;
}

export function applyStructuresToChunk(
  chunk: VoxelChunk,
  structures: ProceduralStructure[],
  structureVoxelDocs: Record<string, any>,
  globalSeed: number
) {
  if (!structures || structures.length === 0) return;

  const radius = 2; // 2x2 chunk anchor radius

  for (let dz = -radius; dz <= radius; dz++) {
    for (let dx = -radius; dx <= radius; dx++) {
      const anchorCX = chunk.cx + dx;
      const anchorCZ = chunk.cz + dz;

      // Deterministic seed for this anchor chunk
      const anchorSeedStr = `${globalSeed}_${anchorCX}_${anchorCZ}`;
      const anchorSeedNum = Array.from(anchorSeedStr).reduce((acc, char) => acc + char.charCodeAt(0), 0);
      const prng = mulberry32(anchorSeedNum);

      for (const structDef of structures) {
        if (structDef.spawnWeight <= 0) continue;

        const roll = prng();
        if (roll > structDef.spawnWeight) continue;

        // Structure spawned!
        const structDoc = structureVoxelDocs[structDef.voxelMapId];
        if (!structDoc || !structDoc.chunks) continue;

        // Decode chunks
        const structChunks = structDoc.chunks;
        
        // Anchor world position
        const anchorWorldX = anchorCX * CHUNK_SIZE_X + Math.floor(CHUNK_SIZE_X / 2);
        const anchorWorldZ = anchorCZ * CHUNK_SIZE_Z + Math.floor(CHUNK_SIZE_Z / 2);
        const anchorWorldY = Math.max(0, structDef.yOffset + 3);

        for (const [ckey, rleData] of Object.entries(structChunks)) {
          // Parse ckey "cx_cz_cy"
          const parts = ckey.split('_').map(Number);
          if (parts.length < 3) continue;
          
          const sCx = parts[0];
          const sCz = parts[1];
          const sCy = parts[2];

          // Decode RLE back to VoxelChunk
          const sChunk = VoxelChunk.deserializePaletteRLEBinary(new Uint8Array(rleData as any));

          for (let ly = 0; ly < CHUNK_SIZE_Y; ly++) {
            for (let lz = 0; lz < CHUNK_SIZE_Z; lz++) {
              for (let lx = 0; lx < CHUNK_SIZE_X; lx++) {
                const idx = VoxelChunk.getIndex(lx, ly, lz);
                const wordLow = sChunk.dataLow[idx];
                const wordHigh = sChunk.dataHigh[idx];

                if (wordLow === VOXEL_WORD_AIR_LOW) continue;

                // World pos relative to structure origin (0,0,0)
                const bwX = (sCx * CHUNK_SIZE_X) + lx;
                const bwY = (sCy * CHUNK_SIZE_Y) + ly;
                const bwZ = (sCz * CHUNK_SIZE_Z) + lz;

                // Offset it by anchor
                const targetWX = anchorWorldX + bwX;
                const targetWY = anchorWorldY + bwY;
                const targetWZ = anchorWorldZ + bwZ;

                // Does it fall into current chunk?
                const targetCX = Math.floor(targetWX / CHUNK_SIZE_X);
                const targetCY = Math.floor(targetWY / CHUNK_SIZE_Y);
                const targetCZ = Math.floor(targetWZ / CHUNK_SIZE_Z);

                if (targetCX === chunk.cx && targetCY === chunk.cy && targetCZ === chunk.cz) {
                  const localX = targetWX - (targetCX * CHUNK_SIZE_X);
                  const localY = targetWY - (targetCY * CHUNK_SIZE_Y);
                  const localZ = targetWZ - (targetCZ * CHUNK_SIZE_Z);

                  if (localX >= 0 && localX < CHUNK_SIZE_X && localY >= 0 && localY < CHUNK_SIZE_Y && localZ >= 0 && localZ < CHUNK_SIZE_Z) {
                    const localIdx = VoxelChunk.getIndex(localX, localY, localZ);
                    chunk.dataLow[localIdx] = wordLow;
                    chunk.dataHigh[localIdx] = wordHigh;
                    chunk.isDirty = true;
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

File: the-lobby/internal/world/structure_placer.go
package world

import (
	"math"
)

// StructurePlacer evaluates multi-chunk structures for a given procedural chunk.
type StructurePlacer struct{}

func NewStructurePlacer() *StructurePlacer {
	return &StructurePlacer{}
}

// PlaceStructures overlays authored VOXEL structures onto the procedural chunk.
func (sp *StructurePlacer) PlaceStructures(chunk *VoxelChunk, def *MapDef, mgr *Manager, globalSeed uint32) {
	if def == nil || len(def.ProceduralStructures) == 0 {
		return
	}

	// For a multi-chunk structure anchor system, we need to check if a structure
	// anchored in a nearby chunk overlaps with the current chunk (chunk.CX, chunk.CY, chunk.CZ).
	// We'll search a 2x2 radius of anchor chunks (from cx-2 to cx+2, cz-2 to cz+2).
	radius := 2

	for dz := -radius; dz <= radius; dz++ {
		for dx := -radius; dx <= radius; dx++ {
			anchorCX := chunk.CX + dx
			anchorCZ := chunk.CZ + dz

			// Deterministic seed for this anchor chunk
			anchorSeed := globalSeed ^ uint32(anchorCX*83123) ^ uint32(anchorCZ*23941)
			prng := NewPRNG(anchorSeed)

			// Roll for each structure
			for _, structDef := range def.ProceduralStructures {
				if structDef.SpawnWeight <= 0 {
					continue
				}

				// Basic check: roll < probability (assume weight is a probability [0.0, 1.0])
				roll := prng.NextFloat()
				if roll > structDef.SpawnWeight {
					continue
				}

				// Structure spawned at anchorCX, anchorCZ!
				// Let's get the voxel blocks from the Manager.
				voxelWorld := mgr.GetVoxelMapBlocks(structDef.VoxelMapID)
				if voxelWorld == nil {
					continue
				}

				// The anchor point inside the anchor chunk (center of chunk)
				anchorWorldX := anchorCX*ChunkSizeX + ChunkSizeX/2
				anchorWorldZ := anchorCZ*ChunkSizeZ + ChunkSizeZ/2
				
				// For now, place it at yOffset + 3 (above bedrock)
				anchorWorldY := structDef.YOffset + 3 
				if anchorWorldY < 0 {
					anchorWorldY = 0
				}

				// Iterate through the loaded voxelWorld and see if any blocks fall into `chunk`.
				voxelWorld.mu.RLock()
				for _, vChunk := range voxelWorld.Chunks {
					for ly := 0; ly < ChunkSizeY; ly++ {
						for lz := 0; lz < ChunkSizeZ; lz++ {
							for lx := 0; lx < ChunkSizeX; lx++ {
								word := vChunk.Get(lx, ly, lz)
								if IsVoxelAir(word) {
									continue
								}

								// World pos of this block relative to the structure's origin (0,0,0)
								bwX := (vChunk.CX * ChunkSizeX) + lx
								bwY := (vChunk.CY * ChunkSizeY) + ly
								bwZ := (vChunk.CZ * ChunkSizeZ) + lz

								// Offset it by the anchor world position
								targetWX := anchorWorldX + bwX
								targetWY := anchorWorldY + bwY
								targetWZ := anchorWorldZ + bwZ

								// Does it fall into the current chunk being generated?
								targetCX := int(math.Floor(float64(targetWX) / float64(ChunkSizeX)))
								targetCY := int(math.Floor(float64(targetWY) / float64(ChunkSizeY)))
								targetCZ := int(math.Floor(float64(targetWZ) / float64(ChunkSizeZ)))

								if targetCX == chunk.CX && targetCY == chunk.CY && targetCZ == chunk.CZ {
									// Yes! Set it in the current chunk.
									localX := targetWX - (targetCX * ChunkSizeX)
									localY := targetWY - (targetCY * ChunkSizeY)
									localZ := targetWZ - (targetCZ * ChunkSizeZ)

									if localX >= 0 && localX < ChunkSizeX && localY >= 0 && localY < ChunkSizeY && localZ >= 0 && localZ < ChunkSizeZ {
										chunk.Set(localX, localY, localZ, word)
									}
								}
							}
						}
					}
				}
				voxelWorld.mu.RUnlock()
			}
		}
	}
}
